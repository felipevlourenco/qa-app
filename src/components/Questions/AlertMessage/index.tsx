import React from 'react'
import styles from 'components/Questions/AlertMessage/styles.module.css'

const AlertMessage = () => {
  return (
    <div className={styles.alert}>
      <span>No questions in the list</span>
    </div>
  )
}

export default AlertMessage
