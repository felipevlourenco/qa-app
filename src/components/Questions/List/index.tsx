import QuestionItem from 'components/Questions/List/Item'
import styles from 'components/Questions/List/styles.module.css'
import { ListProp } from 'components/Questions/List/types'

const List = ({ questions }: ListProp) => {
  return (
    <ul className={styles.list}>
      {questions.map((question) => (
        <QuestionItem key={question.id} question={question} />
      ))}
    </ul>
  )
}

export default List
