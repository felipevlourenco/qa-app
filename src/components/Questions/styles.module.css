.header {
  font-size: 1.2rem;
  font-weight: 600;
}

.row {
  display: flex;
  gap: 0.5rem;
  flex-wrap: wrap;
}

.row > button {
  flex: 1 1 200px;
}