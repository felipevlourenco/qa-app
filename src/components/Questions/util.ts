import { Question } from 'store/questions/types'

const sortQuestion = (a: Question, b: Question) => (a.question.toLowerCase() > b.question.toLowerCase() ? 1 : -1)

export default sortQuestion
