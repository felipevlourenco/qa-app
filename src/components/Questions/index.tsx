import { useMemo } from 'react'
import { useAppSelector, useAppDispatch } from 'hooks/useRedux'
import Button from 'components/Button'
import List from 'components/Questions/List'
import styles from 'components/Questions/styles.module.css'
import useToggle from 'hooks/useToggle'
import sortQuestion from 'components/Questions/util'
import { deleteAllQuestions } from 'store/questions'
import AlertMessage from 'components/Questions/AlertMessage'
import Tooltip from 'components/Tooltip'

const Questions = () => {
  const dispatch = useAppDispatch()
  const { isToggled, toggle } = useToggle()
  const { data: questions } = useAppSelector((questionState) => questionState.questions)

  const list = useMemo(() => {
    return isToggled ? [...questions].sort(sortQuestion) : questions
  }, [isToggled, questions])

  const deleteAllHandler = () => {
    dispatch(deleteAllQuestions())
  }

  return (
    <>
      <Tooltip content="Here you can find the created questions and their answer">
        <span className={styles.header}>Created questions</span>
      </Tooltip>
      {list.length ? <List questions={list} /> : <AlertMessage />}
      <div className={styles.row}>
        <Button onClick={toggle}>Sort questions</Button>
        <Button onClick={deleteAllHandler} variant="secondary">
          Delete all questions
        </Button>
      </div>
    </>
  )
}

export default Questions
