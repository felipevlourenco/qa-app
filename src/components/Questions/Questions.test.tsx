import { fireEvent, render } from '@testing-library/react'
import Questions from 'components/Questions'
import MockStore from 'test/utils/MockStore'

describe('component/Questions', () => {
  it('should render', () => {
    const { container } = render(
      <MockStore>
        <Questions />
      </MockStore>
    )
    expect(container).toMatchSnapshot()
  })

  it('should delete all questions', () => {
    const { getByRole, getByText } = render(
      <MockStore>
        <Questions />
      </MockStore>
    )
    const list = getByRole('list')
    expect(list).toBeInTheDocument()
    const button = getByText('Delete all questions')
    fireEvent.click(button)
    expect(list).not.toBeInTheDocument()
  })

  // it('should sort questions', () => {
  //   const { } = render(
  //     <MockStore
  //       store={{
  //         ...store,
  //         questions: {
  //           ...store.questions,
  //           data: [
  //             ...store.questions.data,
  //             {
  //               id: 2,
  //               question: 'AAAAA',
  //               answer: 'AAAAA'
  //             }
  //           ]
  //         }
  //       }}>
  //       <Questions />
  //     </MockStore>
  //   )
  // })
})
