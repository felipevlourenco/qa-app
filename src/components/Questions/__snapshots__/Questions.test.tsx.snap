// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component/Questions should render 1`] = `
<div>
  <div
    class="tooltip"
  >
    <div
      class="child"
    >
      <span
        class="header"
      >
        Created questions
      </span>
    </div>
  </div>
  <ul
    class="list"
  >
    <li
      class="item"
    >
      <div
        class="button"
      >
        <span
          class="question"
        >
          How to add a question?
        </span>
      </div>
    </li>
  </ul>
  <div
    class="row"
  >
    <button
      class="button primary"
      type="button"
    >
      Sort questions
    </button>
    <button
      class="button secondary"
      type="button"
    >
      Delete all questions
    </button>
  </div>
</div>
`;
